### PR TITLE
Manage version number for builds and releases

### LINKED ISSUE #35: Manage version number for builds and releases
Create managed release number to replace legacy SVN based numbering e.g.:
m: major
n: minor
o: bugfix
p: datetime (yyyymmdd-hhmmss)

p to be added at automatically at buildtime; m,n,o from developer managed version text file.

### PR DESCRIPTION
**Pull request by hsautessella**
This implements a similar a versioning strategy to https://github.com/pace-neutrons/Herbert/pull/108.

- The version (of format `<major>.<minor>.<patch>`), will be held in the top-level file `VERSION`.
- CMake will read this at configure time and output:
    - `horace_get_build_version.m`
    - `version.h`

   which expose the version number to the Matlab and C++ source.
- If the build is a developer build (i.e. CMake is run without the `Horace_RELEASE_TYPE=RELEASE` flag), the current Git revision SHA will be appended to the version.
- If CMake has not yet been run, the Matlab function `horace_version.m` will attempt to read the `VERSION` file directly and will append `.dev` to the end to signify this is an un-built version. If the `VERSION` file cannot be read, `horace_version()` will return `0.0.0.dev`.

The above is also documented in ADR # 11

Fixes #35

### CODE REVIEW

File: _LowLevelCode/cpp/accumulate_cut_c/accumulate_cut_c.cpp [hunk 1]
```diff
@@ -72,10 +73,8 @@ based on Matlab code of T.G.Perring   19 July 2007; C-version Alex Buts 02 July
 
 void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
 {
-
-    const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
     if (nrhs == 0 && nlhs == 1) {
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/accumulate_cut_c/accumulate_cut_c.cpp:76]: `nlhs == 0` needed too

File: _LowLevelCode/cpp/bin_pixels_c/bin_pixels_c.cpp [hunk 2]
```diff
@@ -68,9 +69,8 @@ void mexFunction(int nlhs, mxArray* plhs[], int nrhs, const mxArray* prhs[])
     double* pS, * pErr, * pNpix;   // arrays for the signal, error and number of pixels in a cell (density);
     mxArray* PixelSorted;
     //
-    const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
     if (nrhs == 0 && nlhs == 1) {
-        plhs[0] = mxCreateString(REVISION);
+        plhs[0] = mxCreateString(Horace::VERSION);
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/bin_pixels_c/bin_pixels_c.cpp:73]: ... and `nlhs == 0`

File: _LowLevelCode/cpp/calc_projections_c/calc_projections_c.cpp [hunk 3]
```diff
@@ -83,9 +84,8 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
     double *pEfix(nullptr), k_to_e;
     urangeModes uRange_mode;
 
-    const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
     if (nrhs == 0 && nlhs == 1) {
-        plhs[0] = mxCreateString(REVISION);
+        plhs[0] = mxCreateString(Horace::VERSION);
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/calc_projections_c/calc_projections_c.cpp:88]: ... and `nlhs == 0`

File: _LowLevelCode/cpp/calc_projections_c/calc_projections_c.cpp [hunk 4]
```diff
@@ -377,7 +377,7 @@ void calc_projections_emode(double * const pMinMax,
     firstprivate(nDetectors,nEnergies,ki,urange_mode,emode,singleEfixed,pEfix,pEnergies,k_to_e) //\
     //reduction(min: q1_min,q2_min,q3_min,e_min; max: q1_max,q2_max,q3_max,e_max)
     {
-#pragma omp for 
+#pragma omp for
         for (long ii = 0; ii < nDetectors; ii++)
         {
             //	detdcn=[cosd(det.phi); sind(det.phi).*cosd(det.azim); sind(det.phi).*sind(det.azim)];   % [3 x ndet]
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/calc_projections_c/calc_projections_c.cpp:383]: ?dead code

File: _LowLevelCode/cpp/combine_sqw/combine_sqw.cpp [hunk 5]
```diff
@@ -145,17 +145,15 @@ void combine_sqw(ProgParameters &param, std::vector<sqw_reader> &fileReaders, co
 
 void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
 {
-
-  const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
   if (nrhs == 0 && nlhs == 1) {
-    plhs[0] = mxCreateString(REVISION);
+    plhs[0] = mxCreateString(Horace::VERSION);
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/combine_sqw/combine_sqw.cpp:149]: ... and `nlhs == 0`

File: _LowLevelCode/cpp/hdf_mex_reader/hdf_mex_reader.cpp [hunk 6]
```diff
@@ -2,14 +2,12 @@
...
-
+#include "utility/version.h"
 
 void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
 {
-
-	const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
 	if (nrhs == 0 && nlhs == 1) {
-		plhs[0] = mxCreateString(REVISION);
+		plhs[0] = mxCreateString(Horace::VERSION);
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/hdf_mex_reader/hdf_mex_reader.cpp:10]: ... and `nlhs == 0`

File: _LowLevelCode/cpp/mtimesx_horace/mtimesx_mex.cpp [hunk 7]
```diff
@@ -80,10 +82,8 @@
 
 void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
 {
-
-    const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
     if (nrhs == 0 && nlhs == 1) {
-        plhs[0] = mxCreateString(REVISION);
+        plhs[0] = mxCreateString(Horace::VERSION);
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/mtimesx_horace/mtimesx_mex.cpp:86]: ... and `nlhs == 0`

File: _LowLevelCode/cpp/recompute_bin_data/recompute_bin_data_c.cpp [hunk 8]
```diff
@@ -15,10 +15,8 @@ enum OutputArguments { // unique output arguments,
 
 void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
 {
-
-    const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
     if (nrhs == 0 && nlhs == 1) {
-        plhs[0] = mxCreateString(REVISION);
+        plhs[0] = mxCreateString(Horace::VERSION);
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/recompute_bin_data/recompute_bin_data_c.cpp:19]: ... and `nlhs == 0`

File: _LowLevelCode/cpp/sort_pixels_by_bins/sort_pixels_by_bins.cpp [hunk 9]
```diff
@@ -188,14 +189,13 @@ std::string  verify_index_array(const mxArray * pix_cell_array_ptr, bool &is_int
...
-! 3 -- number of pixels in each cell  (densities) 
+! 3 -- number of pixels in each cell  (densities)
 !
 /**********************************************************************************************/
 void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
 {
-    const char REVISION[] = "$Revision:: 1758 ($Date:: 2019-12-16 18:18:50 +0000 (Mon, 16 Dec 2019) $)";
     if (nrhs == 0 && nlhs == 1) {
-        plhs[0] = mxCreateString(REVISION);
+        plhs[0] = mxCreateString(Horace::VERSION);
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/sort_pixels_by_bins/sort_pixels_by_bins.cpp:198]: ... and `nlhs == 0`

File: horace_core/admin/check_horace_mex.m [hunk 10]
```diff
@@ -1,15 +1,12 @@
-function [rez,n_errors,minVer,maxVer,compilation_date,can_use_mex_4_combine]=check_horace_mex()
+function [rez, n_errors, can_use_mex_4_combine] = check_horace_mex()
 % function checks if horace mex files are compiled correctly and return
 % their SVN min and max version and the latest date these files were
```
Comment (nickbattam-tessella) [inline horace_core/admin/check_horace_mex.m:3]: this comment is now out-of-date

File: horace_core/admin/check_horace_mex.m [hunk 11]
```diff
@@ -21,72 +18,46 @@
 % $Revision: 1738 $    $Date: 2019-03-12 14:31:09 +0000 (Tue, 12 Mar 2019) $
```
Comment (nickbattam-tessella) [inline horace_core/admin/check_horace_mex.m:18]: is removal of the `$Revision ` lines going in a follow-up PR in Herbert and Horace?

File: _LowLevelCode/cpp/combine_sqw/combine_sqw.cpp [hunk 12]
```diff
@@ -1,8 +1,8 @@
-#include "../CommonCode.h"
-
+#include "CommonCode.h"
```
Comment (nickbattam-tessella) [inline _LowLevelCode/cpp/combine_sqw/combine_sqw.cpp:1]: `CommonCode.h` is still in the top `cpp` folder. Is that now on the path so this include will find the file correctly?

Comment (hsautessella) [reply _LowLevelCode/cpp/combine_sqw/combine_sqw.cpp:1]: I've put an `include_directories()` command in `_LowLevelCode/cpp/CMakeLists.txt` so every C++ target will now include the `cpp` directory.

File: horace_core/admin/check_horace_mex.m (hunk 11 above)
Comment (hsautessella) [reply horace_core/admin/check_horace_mex.m:18]: Yep, this was already a big PR

**Pull request merged**